tf2_gnn/layers/graph_densely_connected_transformer.py

```diff
@@ -0,0 +1,219 @@
+        D ~ vertex representation _d_imension
+        H ~ attention _h_ead size
+        N ~ _n_um attention heads
+        P ~ _n_number of node pairs in total, where P = sum(graph_size**2 for graph_size in batch)
+
+        Args:
+            inputs: Tuple of two values:
+                node_embeddings: float Tensor of shape [V, D], containing current representations
+                    of all nodes.
+                in_graph_node_pairs: int Tensor of shape [P, 2], essentially, the set of edges
+                    if all graphs were fully connected.
+            training: Flag indicating if we are in training mode (with dropout) or not.
+        """
+        node_embeddings, in_graph_node_pairs = inputs
+
+        attention_output = self._per_graph_attention_layer(
+            node_embeddings, in_graph_node_pairs, training=training,
+        )  # [V, N * H]
+
+        # Project back to hidden size, drop out, insert residual from input and norm:
+        pre_intermediate_states = self._attention_to_pre_intermediate_layer(
+            attention_output, training=training
+        )
+
+        if training:
+            pre_intermediate_states = tf.nn.dropout(
+                pre_intermediate_states, rate=self._hidden_dropout_rate
+            )
+
+        pre_intermediate_states = self._pre_intermediate_layer_norm(
+            pre_intermediate_states + node_embeddings, training=training
+        )
+
+        # Blow up to intermediate, project back, and drop out:
+        intermediate_states = self._pre_intermediate_to_intermediate_layer(
+            pre_intermediate_states, training=training
+        )
+
+        layer_output = self._intermediate_to_output_layer(
+            intermediate_states, training=training
+        )
+
+        if training:
+            layer_output = tf.nn.dropout(
+                layer_output, rate=self._hidden_dropout_rate
+            )
+
+        # Finally, insert residual from before intermediate blowup, norm and return:
+        layer_output = self._output_layer_norm(
+            layer_output + pre_intermediate_states, training=training
+        )
+
+        return layer_output
+
+    def _per_graph_attention_layer(
+        self,
+        node_embeddings: tf.Tensor,  # [V, D], float
+        in_graph_node_pairs: tf.Tensor,  # [None, 2], int
+        training: bool,
+    ):
+        # Compute queries, keys, values in one go for all heads, then split out:
+        node_qkv = self._node_label_to_qkv_layer(node_embeddings, training=training)
+        node_queries, node_keys, node_values = tf.split(
+            value=tf.reshape(
+                node_qkv, shape=(-1, self._num_heads, 3 * self._head_size)
+            ),
+            num_or_size_splits=3,
+            axis=-1,
+        )  # 3 times [V, N, H]
+
+        # Compute attention scores between queries and keys by doing an inner product,
+        # normalised by the square root of the representation size.
+        scores = tf.einsum(
```

Comment:
I find it useful to keep information about which kind of attention is being implemented when there is an einsum involved in a comment, for example: `Attention(Q, K) = softmax(Q K^T/sqrt(D))` ?

Reply:
Eh, I included the comment above the einsum to clarify the attention kind, but I think what you mean is you would prefer to have it as a math-y formula as well?

Reply:
Yes, I meant the formula, but it's not a biggy. Sorry I forgot about this one - It's easy to forget these PRs exist in amongst all the VSTS noise..